pyccel/ast/cudaext.py

```diff
@@ -0,0 +1,309 @@
+            raise TypeError('Unknown type of  %s.' % type(arg))
+
+        is_homogeneous_tuple = isinstance(arg, (PythonTuple, PythonList, HomogeneousTupleVariable)) and arg.is_homogeneous
+        is_array = isinstance(arg, Variable) and arg.is_ndarray
+
+        # TODO: treat inhomogenous lists and tuples when they have mixed ordering
+        if not (is_homogeneous_tuple or is_array):
+            raise TypeError('we only accept homogeneous arguments')
+
+        # Verify dtype and get precision
+        if dtype is None:
+            dtype = arg.dtype
+            prec = get_final_precision(arg)
+        else:
+            dtype, prec = process_dtype(dtype)
+        # ... Determine ordering
+        order = str(order).strip("\'")
+
+        shape = process_shape(False, arg.shape)
+        rank  = len(shape)
+
+        if rank < 2:
+            order = None
+        else:
+            # ... Determine ordering
+            order = str(order).strip("\'")
+
+            if order not in ('K', 'A', 'C', 'F'):
+                raise ValueError(f"Cannot recognize '{order}' order")
+
+            # TODO [YG, 18.02.2020]: set correct order based on input array
+            if order in ('K', 'A'):
+                order = 'C'
+            # ...
+        #Verify memory location
+        if memory_location not in ('host', 'device', 'managed'):
+            raise ValueError("memory_location must be 'host', 'device' or 'managed'")
+        self._arg   = arg
+        self._shape = shape
+        self._rank  = rank
+        self._dtype = dtype
+        self._order = order
+        self._precision = prec
+        self._memory_location = memory_location
+        super().__init__()
+
+    def __str__(self):
+        return str(self.arg)
+
+    @property
+    def arg(self):
+        return self._arg
+    @property
+    def memory_location(self):
+        return self._memory_location
+
+class CudaSynchronize(PyccelInternalFunction):
+    "Represents a call to  Cuda.deviceSynchronize for code generation."
+
+    __slots__ = ('_dtype','_precision','_shape','_rank','_order')
+    _attribute_nodes = ()
+    def __init__(self):
+        #...
+        self._shape     = None
+        self._rank      = 0
+        self._dtype     = NativeInteger()
+        self._precision = None
+        self._order     = None
+        super().__init__()
+
+class CudaInternalVar(PyccelAstNode):
+    """
+    Represents a General Class For Cuda internal Variables Used To locate Thread In the GPU architecture"
+
+    Parameters
+    ----------
+    dim : NativeInteger
+        Represent the dimension where we want to locate our thread.
+
+    """
+    __slots__ = ('_dim','_dtype','_precision','_shape','_rank','_order')
+    _attribute_nodes = ('_dim',)
+
+    def __init__(self, dim=None):
+        
+        if isinstance(dim, int):
+            dim = LiteralInteger(dim)
+        if not isinstance(dim, LiteralInteger):
+            raise TypeError("dimension need to be an integer")
```

Comment:
I am a little confused by this. According to the docstring this is an internal variable. I am not sure exactly what is meant by this but I assume it is not created by the user. In that case why do we need to use `LiteralInteger` rather than integer?

Reply:
this is just a general class for the internal variables.
which are called in the code as `CudaThreadIdx(x)` where x can be `0, 1, 2`.
i think a check for `PyccelAstNode` would be better.

Note: this will change later as we discussed to using class_def instead. maybe another Pr?

Reply:
I agree that this could be done in another PR, but in that case we need an issue opening which explains the fix clearly

Reply:
That said it may be easier to merge that PR into here to avoid having to write the doc strings for all these unnecessary objects ;P

Reply:
See #16